### ISSUE TITLE
Dataset behaves strangely: symlinks become empty string, send-receive causes 100% CPU hang

### ISSUE THREAD

**Issue by ptommy** [openzfs/zfs#13269]
I think I have a corrupted dataset, and may have had for quite a while without anything detecting it. I realise this may not be possible to solve, but I will describe what I see.

Problem 1: Some symlinks get created pointing to the empty string. I have verified with strace that symlink() is called with correct arguments and returns 0, so `ln` and `rsync` thinks everything's fine, but checking the resulting symlink with `readlink` returns the empty string (""). This only happens for target names longer than 56 chars, but I have existing symlinks with longer target names, created by rsync maybe 1 year ago.

Problem 2: Incremental send-receive creates un-listable directories. The initial non-incremental send of a recent snapshot works fine (except the symlink problem is transfered to the received dataset). lncremental send-receive indicate no errors, but create snapshots in which some random directories make `ls` report Input/Output error, and some other random directories causes `find` to hang on 100% CPU and I can not kill -9 `find` or unmount the dataset so I have to reboot the computer.

I scrub monthly without errors. There's nothing in the logs indicating any problem.

Background: The pool was created  3 years ago. It received 4 datasets from an older pool, and got 9 new created. I only have problems with 1 of the received datasets, so it may have happened during the send-receive, or even before that on the older pool. Some empty symlink targets are present in the oldest 3 year old snapshot, but on the other hand some 1 year old snapshots have symlinks with longer targets that I'm unable to recreate now.

### System information

Type | Version/Name
 --- | ---
Distribution Name	| Gentoo
Distribution Version	| N/A
Kernel Version	| 5.10.103 (with MuQSS patches)
Architecture	| AMD64
OpenZFS Version	| 2.1.2

### Describe how to reproduce the problem
I can reliably reproduce on my system, trying to create symlinks in different datasets, and I can do send-receive, and it's the same directories that get the i/o error / hang every time, although different directories for different snapshots.

**Comment by rincebrain** [comment]
You said before that you only have this problem on one dataset, but then below you say "can reliably reproduce this trying to create symlinks in different datasets".

This sounds like something deeply broken and not necessarily specific to your pool, to me. If you create a fresh pool, are you able to reproduce any of these problems? If you import this pool on a LiveCD of a binary-based distro, does the brokenness behave the same or differently?

**Comment by ptommy** [comment]
When I create symlinks in any other dataset in the pool (as old as the pool or newly created ones) it works fine. Only this one dataset in the pool has this problem, but if I send-receive that dataset to a different pool, the symlink problem persists in the received dataset in the new pool.

I will try your suggestions as soon as I can (some time next week probably).

**Comment by ptommy** [comment]
I tested with a newly created pool today.

- zfs-2.1.2-r1-gentoo
- zfs-kmod-2.1.2-r0-gentoo
- linux 5.10.109-gentoo-x86_64 + MuQSS v0.205

The dataset has some non-default properties, but  I created the target with only default properties, and then tested changing them some at a time. Each test was three different send-receives: an old snap, a young snap, an incremental send from the young snap.

The problem with I/O Error and 100 % CPU hang on directory listings after an incremental send-receive did not go away.

Test 1 was with only default properties on the target. Test 2 was with `xattr=sa acltype=posix`. In both test 1 and 2 there was no symlink problem, and I could create symlinks with long target names in the received dataset. In test 3 I set `dnodesize=auto` in addition to the previous change, and now I got the symlink problem in the received dataset.

I repeated test 3 on the pool on which I first discovered the problems, which is about 3 years old, but with `zpool upgrade` on every new zfs version, and the symlink problem went away there too when I used `zfs receive -o dnodesize=legacy`.

Conclusion so far: If I send-receive my faulty dataset (having `dnodesize=auto`) to a new dataset with `dnodesize=legazy` the link problem (which I have on the original dataset) goes away, but the problem with I/O error and 100 % CPU hang (which I don't have on the original dataset) still manifests in the received dataset after an incremental send-receive.
